### PR TITLE
feat(component): allow delegatesFocus on the component decorator

### PR DESCRIPTION
**Pull request by MrAntix**
ref: https://github.com/ionic-team/stencil/issues/1008

### CODE REVIEW

**Comment by adamdbradley** [comment]
I like this, but wonder if we should instead have the options object passed in, rather than making a new option to the component decorator.

**Comment by manucorporat** [comment]
Agreed with Adam, i think it should be an object, probably allowing the `shadow` property to be both boolean or `ShadowDomOptions`:

```
 @Component({
     shadow: {
         delegateFocus: true
     }
  })
```

**Comment by adamdbradley** [comment]
@MrAntix would you be able to make the changes so `shadow` could either be a boolean, or an object? If it's an object we merge it with `{ mode: 'open' }`. And if it's a boolean, the object should be `{ mode: 'open' }`.

**Comment by MrAntix** [comment]
hows that look?

**Comment by adamdbradley** [comment]
Looking great. One thing is that in most cases, `shadow: true` will be the case, and the options would be used less often. So I think the net result of less could we be to not always build an options object, where every component would set `{mode:'open'}`, but rather have the run do the merge instead.

File: src/core/host-snapshot.ts [hunk 1]
```diff
@@ -55,7 +55,7 @@ export function initHostSnapshot(domApi: d.DomApi, cmpMeta: d.ComponentMeta, hos
       // this component is using shadow dom
       // and this browser supports shadow dom
       // add the read-only property "shadowRoot" to the host element
-      domApi.$attachShadow(hostElm, { mode: 'open' });
+      domApi.$attachShadow(hostElm, cmpMeta.shadowRootInit);
```
Comment (manucorporat) [inline src/core/host-snapshot.ts]:
```suggestion
      domApi.$attachShadow(hostElm, { mode: 'open', cmpMeta.shadowRootInit });
```

File: src/declarations/component.ts [hunk 2]
```diff
@@ -28,13 +28,14 @@ export interface ComponentMeta {
...
   encapsulationMeta?: number;
+  shadowRootInit?: Readonly<ShadowRootInit>;
   assetsDirsMeta?: AssetsMeta[];
   componentConstructor?: ComponentConstructor;
   componentClass?: string;
   dependencies?: ComponentDependencies;
   jsdoc?: JsDoc;
   styleDocs?: StyleDoc[];
-  hmrLoad?: () => void;
+  hmrLoad?: () => void;
```
Comment (manucorporat) [inline src/declarations/component.ts]: revert

File: src/compiler/transpile/datacollection/component-decorator.ts [hunk 3]
```diff
@@ -107,3 +116,7 @@ function getHostMeta(diagnostics: d.Diagnostic[], hostData: d.HostMeta) {
 
   return hostData;
 }
+
+function isShadowDomOptions(o: any): o is d.ShadowDomOptions {
+  return typeof(o) != typeof(true);
```
Comment (manucorporat) [inline src/compiler/transpile/datacollection/component-decorator.ts]:
```suggestion
  return o && typeof(o) === 'object';
```

File: src/compiler/transpile/datacollection/component-decorator.ts [hunk 4]
```diff
@@ -47,14 +46,24 @@ It will be removed in future versions. Please use the "hostData()" method instea
...
+    jsdoc: serializeSymbol(checker, symbol)    
   };
 
-  // normalizeEncapsulation
-  cmpMeta.encapsulationMeta =
-      componentOptions.shadow ? ENCAPSULATION.ShadowDom :
+  // shadow dom options
+  if(componentOptions.shadow){
+    cmpMeta.encapsulationMeta = ENCAPSULATION.ShadowDom
+    if(isShadowDomOptions(componentOptions.shadow)) {
```
Comment (manucorporat) [inline src/compiler/transpile/datacollection/component-decorator.ts]:
```suggestion
    if(isShadowDomOptions(componentOptions.shadow)) {
        cmpMeta.shadowRootInit = componentOptions.shadow;
```

Comment (manucorporat) [reply src/compiler/transpile/datacollection/component-decorator.ts]: remove the DEFAULT_SHADOW_ROOT_INIT from here, it should be provided by the runtime automatically

File: src/compiler/transpile/datacollection/component-decorator.ts [hunk 5]
```diff
@@ -47,14 +46,24 @@ It will be removed in future versions. Please use the "hostData()" method instea
...
+      cmpMeta.shadowRootInit = Object.assign({}, DEFAULT_SHADOW_ROOT_INIT, {
+        delegatesFocus: componentOptions.shadow.delegatesFocus
+      });
+    }else {
+      cmpMeta.shadowRootInit = DEFAULT_SHADOW_ROOT_INIT;
```
Comment (manucorporat) [inline src/compiler/transpile/datacollection/component-decorator.ts]: not needed

File: src/compiler/transpile/datacollection/component-decorator.ts [hunk 6]
```diff
@@ -47,14 +46,24 @@ It will be removed in future versions. Please use the "hostData()" method instea
     assetsDirsMeta: [],
     hostMeta: getHostMeta(diagnostics, componentOptions.host),
     dependencies: [],
-    jsdoc: serializeSymbol(checker, symbol)
+    jsdoc: serializeSymbol(checker, symbol)
```
Comment (manucorporat) [inline src/compiler/transpile/datacollection/component-decorator.ts]: revert

**Comment by MrAntix** [comment]
all done, do you need the commits squashed?

**Comment by manucorporat** [comment]
@MrAntix don't worry about it! we can squash them before merge

@adamdbradley, @jthoms1 the PR looks good to me!

**Comment by tehdb** [comment]
please merge and publish 🙏

**Comment by zoofadoofa** [comment]
This is not really an edge case, anything that is slotting in content into shareable components will need delegatesFocus
[See Demo](https://jsfiddle.net/zoofadoofa/6hvrzkf4/)

**Comment by MrAntix** [comment]
Could this be released? I won't have capacity to upgrade to v1 as the breaking changes are too big and my project is going live with 0.18.x

**Comment by romulocintra** [comment]
Any news ;) any help needed ?

**Comment by driskull** [comment]
This would be nice to have

**Comment by adamdbradley** [comment]
Closing in favor of https://github.com/ionic-team/stencil/pull/1190

**Comment by MrAntix** [comment]
Thanks

**Pull request closed without merging**